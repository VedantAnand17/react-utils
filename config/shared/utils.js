/**
 * Misc utilities reusable in different JS-format config files.
 */
/* eslint-disable import/no-extraneous-dependencies */

const fs = require('fs');
const path = require('path');
const cssesc = require('cssesc');

const { interpolateName } = require('loader-utils');

/**
 * Normalizes file path to OS-independent format (adopted from css-loader).
 * @ignore
 * @param {string} file
 * @return {string}
 */
function normalizePath(file) {
  return path.sep === '\\' ? file.replace(/\\/g, '/') : file;
}

// eslint-disable-next-line no-control-regex
const filenameReservedRegex = /[<>:"/\\|?*]/g;
// eslint-disable-next-line no-control-regex
const reControlChars = /[\u0000-\u001f\u0080-\u009f]/g;

function escapeLocalident(localident) {
  return cssesc(
    localident
      // For `[hash]` placeholder
      .replace(/^((-?[0-9])|--)/, '_$1')
      .replace(filenameReservedRegex, '-')
      .replace(reControlChars, '-')
      .replace(/\./g, '-'),
    { isIdentifier: true },
  );
}

/**
 * Returns the name of package containing the folder; i.e. it recursively looks
 * up from the folder for the closest package.json file, and returns the name in
 * that file. It also caches the results from previously fisited folders.
 * @ignore
 * @param {string} folder
 * @return {string}
 */
function getPackageInfo(folder) {
  let res = getPackageInfo.cache[folder];
  if (!res) {
    const p = path.resolve(folder, 'package.json');
    /* eslint-disable global-require, import/no-dynamic-require */
    res = fs.existsSync(p) ? {
      name: require(p).name,
      root: folder,
    } : getPackageInfo(path.resolve(folder, '..'));
    /* eslint-enable global-require, import/no-dynamic-require */
    getPackageInfo.cache[folder] = res;
  }
  return res;
}
getPackageInfo.cache = {};

function getLocalIdent(
  { resourcePath },
  localIdentName,
  localName,
  options = {},
) {
  const packageInfo = getPackageInfo(path.dirname(resourcePath));
  const request = normalizePath(path.relative(packageInfo.root, resourcePath));
  return interpolateName({
    resourcePath,
  }, localIdentName, {
    ...options,
    content: `${packageInfo.name + request}\x00${localName}`,
    context: packageInfo.root,
  }).replace(/\[package\]/gi, packageInfo.name)
    .replace(/\[local\]/gi, localName)
    .replace(/@/g, '-');
}

function generateScopedNameDev(localName, assetPath) {
  // Mind: "css-loader" additionally escapes identifiers returned by
  // getLocalIdent(); the babel plugins - don't, thus we need to escape
  // ourselves.
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[package]___[path][name]___[local]___[hash:base64:6]',
      localName,
      {},
    ),
  );
}

function generateScopedNameProd(localName, assetPath) {
  return escapeLocalident(
    getLocalIdent(
      { resourcePath: assetPath },
      '[hash:base64:6]',
      localName,
      {},
    ),
  );
}

module.exports = {
  generateScopedNameDev,
  generateScopedNameProd,
  getLocalIdent,
};
