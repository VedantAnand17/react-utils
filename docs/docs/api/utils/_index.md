- [api](/docs/api/utils/api) - An alias for [axios] library (HTTP(S) client).
- [config](/docs/api/utils/config) - Isomorphic app config.
- [isomorphy](/docs/api/utils/isomorphy) - Helpers for isomorphic code logic.
- [Jest Utils](/docs/api/utils/jest-utils) &mdash; Helpers useful in [Jest] tests.
- [PT (prop-types)](/docs/api/utils/prop-types) - An alias for [prop-types] lib
  for runtime ReactJS prop type checks.
- [React Global State](/docs/api/utils/react-global-state) - Aliases for
  [react-global-state](https://dr.pogodin.studio/docs/react-global-state/index.html)
  lib.
- [server](/docs/api/utils/server) - Server-side constants and functions.
- [themed](/docs/api/utils/themed) - Aliases for
  [React Themes](https://dr.pogodin.studio/docs/react-themes) library.
- [time](/docs/api/utils/time) - Date & time utils.
- [webpack](/docs/api/utils/webpack) - [Webpack]-related utils.

<!-- links -->
[axios]: https://www.npmjs.com/package/axios
[ESLint]: https://eslint.org
[Jest]: https://jestjs.io
[prop-types]: https://www.npmjs.com/package/prop-types
[Stylelint]: https://stylelint.io
[Webpack]: https://webpack.js.org
