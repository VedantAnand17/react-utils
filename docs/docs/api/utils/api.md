---
sidebar_class_name: deprecated
---
# api
[api]: /docs/api/utils/api
:::danger Deprecated
The [api] alias for [axios] library has been removed in **v1.39.0**
because it turned out incompatible with the new [mockAxios()] helper
in [Jest Utils] (it was causing a circular dependency in some tests),
and it provided little added value in general. Use [axios] directly instead.
:::

```js
import { api } from '@dr.pogodin/react-utils';
```
The [api] is just an alias for [axios]
library, which provides an isomorphic HTTP(S) client.

[axios]: https://www.npmjs.com/package/axios
[Jest Utils]: /docs/api/utils/jest-utils
[mockAxios()]: /docs/api/utils/jest-utils#mockaxios
