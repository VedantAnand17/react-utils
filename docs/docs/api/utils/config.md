# config
```js
import { config } from '@dr.pogodin/react-utils';
```
The `config` module provides isomorphic application configuration, based on
the [node-config] library. At the server-side it works just as the [node-config]
library import. At the client side  it is makes the config available in the same
format _via_ config injection into HTML template during the server-side
rendering.

These are two special things about the client-side config:
- Any configuration in the `SECRET` config segment is stripped out before
  the injection into the frontend template by the server-side renderer, thus
  providing a configuration segment where server-side only secrets can be
  stored.

  **Example**
  ```js title="/config/default.json"
  module.exports = {
    // These configs are available both client- and server-side:
    SAMPLE_KEY: 'sample value',
    SAMPLE_SEGMENT: {
      KEY_1: 'value 1',
      KEY_2: 'value 2',
    },
    SAMPLE CONFIG_KEY_1: 'VALUE',
    CO
  };
  ```
- Some extra fields required for internal workings of `react-utils` library
  may be added to the config.

[node-config]: https://github.com/lorenwest/node-config
