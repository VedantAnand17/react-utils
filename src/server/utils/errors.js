// Assertion & error handling helpers.

import {
  StatusCodes as CODES,
  ReasonPhrases as ERRORS,
  getReasonPhrase as getErrorForCode,
} from 'http-status-codes';

import joi from 'joi';

/**
 * Auxiliary aliases.
 */
export {
  CODES,
  ERRORS,
  getErrorForCode,
  joi,
};

/**
 * Creates a new error with given message, and attached HTTP status code.
 * @param {string} message Error message.
 * @param {number} [statusCode=500] Optional. HTTP status code. Defaults to 500
 *  (Internal Server Error).
 * @return {Error}
 */
export function newError(message, statusCode = CODES.INTERNAL_SERVER_ERROR) {
  const error = new Error(message);
  error.status = statusCode;
  return error;
}

/**
 * Throws an error with given message and HTTP status code.
 * @param {string} message Error message.
 * @param {number} [statusCode=500] Optional. HTTP error code. Defaults to 500
 *  (Internal Server Error).
 * @throws {Error}
 */
export function fail(message, statusCode = CODES.INTERNAL_SERVER_ERROR) {
  throw newError(message, statusCode);
}

/**
 * Validates a value using the given Joi schema, and throws an error with
 * the given message and HTTP status code in case of the validation failure.
 * @param {any} value
 * @param {object} schema
 * @param {string} [message] Optional. Error message.
 * @param {number} [statusCode=500] Optional. HTTP status code. Defaults to 400
 *  (Bad Request).
 * @throws {Error}
 */
export function assert(
  value,
  schema,
  message = '',
  statusCode = CODES.BAD_REQUEST,
) {
  const { error } = schema.validate(value, { abortEarly: false });
  if (error) {
    fail(message.concat(message ? '\n' : '', error.message), statusCode);
  }
}
