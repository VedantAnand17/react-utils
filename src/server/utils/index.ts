// Aggregates utility modules together.

export * as errors from './errors';
