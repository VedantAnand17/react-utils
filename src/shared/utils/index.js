import path from 'path';

import config from './config';
import * as isomorphy from './isomorphy';
import * as redux from './redux';
import * as webpack from './webpack';

const juUrl = module.webpackPolyfill ? './shared/utils/jest'
  : path.resolve(__dirname, './jest');

const JU = isomorphy.isServerSide() ? webpack.requireWeak(juUrl) : null;

export {
  config,
  isomorphy,
  JU,
  redux,
  webpack,
};
