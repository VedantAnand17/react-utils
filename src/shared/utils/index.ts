import themedImpl, {
  type Theme,
  COMPOSE,
  PRIORITY,
  ThemeProvider,
} from '@dr.pogodin/react-themes';

import config from './config';
import * as isomorphy from './isomorphy';
import time from './time';
import * as webpack from './webpack';

import type * as JuT from './jest';

export {
  Barrier,
  Emitter,
  Semaphore,
  withRetries,
} from '@dr.pogodin/js-utils';

export { getSsrContext } from './globalState';
export { default as splitComponent } from './splitComponent';

type ThemedT = typeof themedImpl & {
  COMPOSE: typeof COMPOSE;
  PRIORITY: typeof PRIORITY;
};

const themed: ThemedT = themedImpl as ThemedT;

themed.COMPOSE = COMPOSE;
themed.PRIORITY = PRIORITY;

// Note: it should be done this way, as in some environments
// "process" might not exist, and process.env.NODE_CONFIG_ENV
// not injected by Webpack.
let NODE_CONFIG_ENV;
try {
  NODE_CONFIG_ENV = process.env.NODE_CONFIG_ENV;
} catch { /* noop */ }

const env = NODE_CONFIG_ENV || process.env.NODE_ENV;

const JU: typeof JuT | null = env !== 'production'
  ? webpack.requireWeak('./jest', __dirname) as typeof JuT | null
  : null;

export {
  type Theme,
  config,
  isomorphy,
  JU,
  themed,
  ThemeProvider,
  time,
  webpack,
};
