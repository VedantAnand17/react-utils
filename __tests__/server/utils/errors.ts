import { fail } from 'server/utils/errors';

test('fail()', () => {
  let error: any;
  let never: never;
  try {
    never = fail('Mock error', 418);
  } catch (e: any) {
    error = e;
  }
  expect(error).toMatchSnapshot();
  expect(error.status).toBe(418);

  // @ts-expect-error
  expect(never).toBeUndefined();
});
