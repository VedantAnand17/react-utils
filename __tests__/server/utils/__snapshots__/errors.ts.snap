// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fail() 1`] = `[Error: Mock error]`;
